### ISSUE TITLE
Overview of Plots.jl Chapter

### ISSUE THREAD

**Issue by storopoli** [JuliaDataScience/JuliaDataScience#36]
Here is an opinionated version. Feel free to criticize:

Subsections for Plots

Brief overview of the `JuliaPlots` ecosystem 

	- `JuliaPlots` Organization
	- Plots.jl (what is it good for and what are its limitations)
	- StatsPlots.jl (what is it good for and what are its limitations)
	- Makie.jl (what is it good for and what are its limitations)
	- AlgebraOfGraphics.jl (what is it good for and what are its limitations)

1. What is Plots.jl?
	- `plot` vs `plot!`
	- input data
	- Series Types and functions(e.g. `line`, `line!`, `heatmap`, `heatmap!`)
	- How to save a plot

2. Backends (I would cover only 3, advantages vs disadvantages):
	- `GR` for everyday stuff
	- `PGFPlotsX` (not the PGFPlots since it's being replaced and deprecated) for publication
	- `Plotly` for interactivity

3. Attributes
	- Overview (What are attributes, and the whole symbol e.g. `:xticks` system)
	- Series attributes
	- Plot Attributes
	- Footnote about the `extra_kwargs` stuff
	- Examples using the most common things that you want to do in a visualization. This could be inserted right after you introduce a specific attribute.

4. Color and Palettes
	I think we should cover colorbrewer and the ones from matplotlib (`inferno`, `viridis`, `magma`). 
	We should cover some stuff from the [Claus Wilke Fundamentals of Data Visualization book](https://clauswilke.com/dataviz/color-basics.html) (Chapters 4 and 19). Also we should cover the three types of color usage:
	1. sequential: continuous stuff, e.g. `:blues` (only blue)
	2. diverging: continuous stuff, e.g. `:RdBu` (from red to blue)
	3. distinguishable: discrete stuff, e.g. `:Set1_5`
	
	I have a very strong positive bias towards colorbrewer `Set`s (e.g. `palette=:Set1_5`).
	
	We should also mention that the reader should use a colorblinded-friendly palette or colors. Maybe we should include an official statistics regarding the prevalence of any sort of colorblindness or color difficulties in the population. I remember seeing somewhere that it was around 5% of people.

5. Layouts
	- Overview on several ways to do layouts
	- the `layout` argument, also cover the `grid`
	- the `@layout` macro
	- specific measures with the `Plots.PlotMeasures` submodule (additionally a footnote to say that we have it in `StatsPlots.PlotMeasures` also)
	- adding subplots incremententally. Define `p1`, `p2`, `p3`; then do a `plot(p1, p2, p3; layout=l)`

6. Animations (?): should we include?
7. Themes (?): should we include? Or can we cover everything in the Attributes subsection?

> OBS1: we are only covering official JuliaPlots packages and backends. So, no `Gadfly`, actually @rikhuijzer had a terrible experience trying to contribute to `Gadfly`...

> OBS2: We are using the [blue codestyle](https://github.com/invenia/BlueStyle)

> OBS3: @rikhuijzer has some opinions on the pattern below. He thinks it is bad and we should not promote.
   ```julia
   p = plot()
   for ...
       ...
       plot!(p, ...)
   end
   ```

>OBS4: We are totally against pie charts. Humans can't distinguish angles very good and it is always better to stick to propotions (bar plot). I will fish out a reference for this, but we have a whole scientific area that studies how humans process visual information. #banpiecharts.

**Comment by lazarusA** [comment]
Now that think about this, what do you think about having just one chapter?  Data Visualisation with subsection for each plotting library, as in 
# Data Visualisation 
Here it goes,
Brief overview of the JuliaPlots ecosystem
## Plots.jl
1. What is Plots.jl? (I can do this one)
using Plots; gr() # here from the start using the default backend
2. Backends  *(could some else please do this one)*
additional ones
3. Attributes (I can do this one)
4. Color and Palettes (I can do this one)
5. Layouts (I can do this one)
6. Animations (yes )  *(could some else please do this one)*
7. Themes (not sure also)  *(could some else please do this one)*
##  StatsPlots *(could some else please do this one)*
..
..

## Makie.jl (I will do all this)
1. What is Makie? 
2. Backends 
Descriptions 
- CairoMakie
- GLMakie
3. plot functions, lines, scatter, image, heatmap, etc...
4.  Attributes
4. Color and Palettes
5. Layouts
6. Animations (yes )
7. Themes (yes)

## AlgegraOfGraphics (I will do all this, suggestions about most needed things using DataFrames are welcome)
.
.
.

**Comment by storopoli** [comment]
My 2 cents on StatsPlots

# StatsPlots

1. A brief intro and show that is is really just "syntactic sugar" for Plots.jl. Also note that works on all plots.jl stuff:
	- Backends
	- Attributes
	- Colours
	- Layouts
	- Animations
	- Themes

2. Works on Types of `Distribution`s and `DataFrame`s, but we will only focus on `DataFrame`s

3. the `@df` macro and the `Symbol`s (`:col1`, `:col2`) instead of other input data (vectors, arrays, tuples etc.)

4. Recipes (I would cover almost all of them, except for `andrewsplot` MDS plot, Dendograms, QQ-Plot etc., since they are specific for modeling or simulation):
	- `histogram`, `histogram2d` and `ea_histogram`
	- `groupedhist` and `groupedbar`
	- `boxplot`
	- `dotplot`
	- `violin`
	- `marginalhist`, `marginalkde` and `marginalscatter`
	- `corrplot` and `cornerplot`

**Comment by lazarusA** [comment]
Good.

- animations (I can also help, we can do it all-together)
- Themes, I not to familiar with these in plots but sure let's do them also, because in Makie definitely they are a must. 

ok. Then we have a clear plan now.

**Comment by rikhuijzer** [comment]
> Now that I think about this, what do you think about having just one chapter?

What would be the arguments in favor of this? But, I don't mind. We can always convert it back to multiple chapters again. You are the expert on plotting, @lazarusA, in my opinion so you probably know best :)

**Comment by storopoli** [comment]
also @rikhuijzer has a very good experience with `AlgebraOfGraphics`, so he might chip in also...

**Comment by rikhuijzer** [comment]
Yes, I would love to do the AlgebraOfGraphics chapter!! I'm really enjoying working with that

**Comment by lazarusA** [comment]
@rikhuijzer my main motivation for a complete chapter is  to not distract some readers with extra chapters. So that they can say, ok just go to chapter X. You will find what you need there, instead of saying go to chapter x1 or x2 or x3, so probably there. 

Ok. Then lets do this. @rikhuijzer you can start with AlgebraOfGraphics, and I will do Makie and the sections mentioned in plots. It will be good to upload the complete template into the main branch, so that we work in our respective sections, but having the general context. @rikhuijzer (or @storopoli)  could you do it? Otherwise,  I can do it on Wednesday afternoon.

**Comment by athulsudheesh** [comment]
My suggestion would be to concentrate on just "one ecosystem". Either go with Plots or Makie. The more options we throw at someone who's new to the language & data science, the more overwhelmed they get with the material.

**Comment by rikhuijzer** [comment]
In an earlier version, Jose and I were thinking about separate chapters and telling the reader to choose one of the plotting libraries. What do you think of that, @athulsudheesh?

**Comment by storopoli** [comment]
We think that covering both ecosystems are good we can add a NOTE warning at the beginning of the section telling the reader that he can do just fine with just one ecosystem.

**Comment by lazarusA** [comment]
I'm fine also with the Chapter structure, but I think is worthwhile to do it for more than one library. And, it's not like students read everything sequentially. In my experience they just go for the thing that looks more similar to their problem and copy paste that example. And then they just adapt to their needs.
